### ISSUE TITLE
loss.backward() inside SimpleLossCompute

### ISSUE THREAD

**Issue by xuefei1** [harvardnlp/annotated-transformer#14]
In the `__call__` function of class `SimpleLossCompute`, `loss.backward()` is called even if `opt` is None, i.e. during validation. Wouldn't this cause gradients to be computed on the validation set and thus you are training on the validation data?

**Comment by gussmith** [comment]
Yes, it appears that you are correct. 

The original code calculates gradients during validation.
I verified that this _trains_ the model towards the validation data, which should not happen. 

With the code below, the model improves quite faster on the training data set.
One major advantage is that with the changes below, the code goes through the validation set twice as fast as before because it does not do `loss.backward()`. 

The recommended way to run validation is: 
```
model.eval()
torch.no_grad()
```
[https://discuss.pytorch.org/t/model-eval-vs-with-torch-no-grad/19615](url)
https://discuss.pytorch.org/t/two-small-questions-about-with-torch-no-grad/27571

And indeed, the code that runs without error is the following: 
```
        model.eval()
        with torch.no_grad():
            test_loss = run_epoch(val_dataloader, model, 
                  SimpleLossCompute(model.generator, criterion, None))
            print("test_loss", test_loss)
```
and: 
```
class SimpleLossCompute:
    "A simple loss compute and train function."
    def __init__(self, generator, criterion, opt=None):
        self.generator = generator
        self.criterion = criterion
        self.opt = opt
        
    def __call__(self, x, y, norm):
        x = self.generator(x)
        loss = self.criterion(x.contiguous().view(-1, x.size(-1)), 
                              y.contiguous().view(-1)) / norm
        #loss.backward()
        if self.opt is not None:
            loss.backward()
            self.opt.step()
            self.opt.optimizer.zero_grad()
        return loss.data * norm
```

**Comment by mathematicsofpaul** [comment]
@gussmith hey gus, did you ever get this code to work? interested in comparing answers?
